File: cs101/pa1/List.java
//Kevin Lee
//1480757
//pa1
//cmps101
//List.java

class List{
    private class Node{
        int data;
        Node next;
        Node previous;

        Node(int data){
            this.data = data; 
            next = null;
            previous = null; 
        }

        public String toString(){ 
            return String.valueOf(data); 
        } 
    }

    private Node front;
    private Node back; 
    private int length;
    private int indexCursor; 
    private Node cursor;

    List(){
        indexCursor = 0;
        front = null;
        back = null;
        cursor = null; 
        length = 0;
    }

    int length(){
        return length;
    }

    int index(){
        if (indexCursor >= -1){
            return indexCursor;
        }
        else{ return -1;
        }
    }

    int front(){ 
        if (length()>0)
            return front.data;
        else
            throw new RuntimeException("Cannot front() on empty list.");
    }

    int back(){
        if (length()>0)
            return back.data;
        else 
            throw new RuntimeException("Cannot back() on empty list.");
    }

    int get(){
        if (length()>=0 && indexCursor>=0)
            return cursor.data;
        else 
            throw new RuntimeException("Cannot get() on empty list.");
    }

    boolean equals(List L){
        boolean bool = true;
        Node A = this.front;
        Node B =L.front;
        if (length == L.length){
            while(bool ==true && A!=null){
                bool =(A.data == B.data);
                A= A.next;
                B= B.next;
            }
            return bool;
        }
        return false;
    }  

    void clear(){
        front = null;
        back = null;
        indexCursor= 0;
        length = 0;
    } 

    void moveFront(){
        if (length>0){
            cursor = front;
            indexCursor = 0;
        }else{ return;
        }
    }

    void moveBack(){
        if(length > 0){
            cursor = back;
            indexCursor = length-1;
        }else{ return;
        }
    }

    void movePrev(){
        if (cursor !=null && indexCursor != 0){
            cursor = cursor.previous;
            indexCursor--;
        } 
        else if (indexCursor== 0){
            indexCursor = -1;
            cursor= null;

        }else{
            return;
        }
    }

    void moveNext(){
        if (cursor!= null && cursor!= back){
            cursor = cursor.next;
            indexCursor++;
        }else if (cursor!= null && cursor == back){
            cursor =null;
            indexCursor=-1;
        }
        else{
            return;
        }
    }

    void prepend(int data){
        Node element = new Node(data);
        if (length==0){
            front = back = element;
            length++;
        }
        else{
            element.next =front;
            front.previous = element;
            front = element;
	    indexCursor++;
            length++;
        }
    }

    void append(int data){
        Node element = new Node(data);  
        if(length==0){
            front = back = element;
            length++;
        }
        else{
            back.next = element;
            element.previous = back;
            back = element;
            length++;
        }
    }

    void insertBefore(int data){
        Node element = new Node(data);
        Node previous = cursor.previous;
        if( length <= 0 || indexCursor< 0){
            throw new RuntimeException( "List Error: insertAfter called on an empty list.");
        }
        if( indexCursor == 0){
            prepend(data);
        }
        else{
            element.previous = previous;
            element.next = cursor;
            previous.next = element;
            cursor.previous= element;
            length++;
            indexCursor++;
        }
    }

    void insertAfter(int data){
        Node element = new Node(data);
        Node next = cursor.next;
        if( length <= 0 || indexCursor < 0){
            throw new RuntimeException( "List Error: insertAfter called on an empty list.");
        }
        if( indexCursor == length - 1){
            append(data);
        }

        else{
            element.previous = cursor;
            element.next = cursor.next;
            next.previous = element;
            cursor.next= element;
            length++;
            indexCursor++;
        }
    }

    void deleteFront() {
        if(length <= 0){
            throw new RuntimeException("List Error: deleteFront called on an empty list.");
        }
        else{
            front = front.next;
            length--;
        }
    }  

    void deleteBack(){
        if(length <= 0){
            throw new RuntimeException("List Error: deleteBack called on an empty list.");
        }
        else{
            cursor = back;
            back = back.previous;
            length--;
        }
    } 

    void delete(){
        if(length() <= 0 || indexCursor< 0){
				throw new RuntimeException("List Error: delete called on an empty list.");
			}
			else{
				cursor.previous.next = cursor.next;
				cursor.next = cursor.previous;
				cursor = null;
				indexCursor = -1;
				length--;
			}
		}

    public String toString(){
        String str = "";
        for(Node N=front; N!=null; N=N.next){
            str += N.toString() + " ";
        }
        return str;
    }

    List copy() {
        List newList = new List();
        Node temp = front;
        while(temp != null){
            newList.append( temp.data );
            temp = temp.next;
        }

        return newList;
    }

}

File: cs101/pa2/List.c
//Kevin Lee
//1480757
//pa1
//cmps101
//List.java

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

typedef struct NodeObj{
        int data;
        struct NodeObj* next;
        struct NodeObj* previous;
    } NodeObj;

typedef struct NodeObj* Node;

typedef struct ListObj{
    Node front;
    Node back;
    Node cursor;
    int indexCursor;
    int length;
    } ListObj;
typedef ListObj* List;

 Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   N->previous = NULL;
   return(N);
}

List newList(void){
   List L;
   L = malloc(sizeof(ListObj));
   L->front = NULL;
   L->front= NULL;
   L->cursor = NULL;
   L->indexCursor = -1;
   L->length = 0;
   return(L);
}

void freeList(List* pL){
    if( *pL!=NULL && pL!=NULL ){
        clear(*pL);
        free(*pL);
        *pL = NULL;
    }
}

void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}


int length(List L){
        return (L->length);
    }

int index(List L){
        if (L->indexCursor >= -1){
            return L->indexCursor;
        }
        else{ return -1;
        }
}

int front(List L){
        if (L->length>0)
            return L->front->data;
        else
            printf("List Error: cannot call front() on empty List\n");
 	     exit(1);
    }

int back(List L){
        if (length(L)>0)
            return L->back->data;
        else
            printf("List Error: cannot call front() on empty List\n");
             exit(1);
    }

int get(List L){
        if (length(L)>=0 && L->indexCursor>=0)
            return L->cursor->data;
        else
            printf("List Error: cannot call front() on empty List\n");
             exit(1);
    }

int equals(List A, List B){
        int  equal = ( A->length == B->length );
        Node C = A->front;
        Node D = B->front;
        if (A->length == B->length){
             while( equal && C!=NULL){
                equal =(C->data == D->data);
                C= C->next;
                D= D->next;
            }
            return equal;
        }
	return equal;
  }

void clear(List L){
        if( L==NULL ){
        printf("List Error: calling clear() on NULL List reference");
        exit(1);
    }
    while( L->front!=NULL ){
        deleteFront(L);
    }
    L->front = NULL;
    L->back = NULL;
    L->cursor = NULL;
    L->length = 0;
    L->indexCursor = -1;
}
void moveFront(List L){
        if (L->length>0){
            L->cursor = L->front;
            L->indexCursor = 0;
        }else{ return;
        }
    }

void moveBack(List L){
        if(L->length > 0){
            L->cursor = L->back;
            L->indexCursor = L->length-1;
        }else{ return;
        }
    }

void movePrev(List L){
        if (L->cursor !=NULL && L->indexCursor!= 0){
            L->cursor = L->cursor->previous;
            L->indexCursor--;
        }
        else if (L->indexCursor== 0){
            L->indexCursor = -1;
            L->cursor= NULL;

        }else{
            return;
        }
    }

 void moveNext(List L){
        if (L->cursor!= NULL && L->cursor!=L->back){
        L->cursor = L->cursor->next;
            L->indexCursor++;
        }else if (L->cursor!= NULL && L->cursor == L->back){
            L->cursor =NULL;
            L->indexCursor=-1;
        }
        else{
            return;
        }
    }

void prepend(List L, int data){
    if( L==NULL ){
        printf("List Error: calling prepend() on NULL List reference");
        exit(1);
    }
    Node element = newNode(data);
    if( L->length==0 ){
        L->front = element;
        L->back = element;
    }else{
        L->front->previous = element;
        element->next = L->front;
        L->front = element;
        L->indexCursor++;
    }
   L->length++;
}

void append(List L, int data){
        Node element = newNode(data);
    if( L->length==0 ){
        L->front = element;
        L->back = element;
	L->length++;
    }else{
        L->back->next = element;
        element->previous = L->back;
        L->back = element;
	L->length++;
    }
}

void insertBefore(List L, int data){
        Node element = newNode(data);
        Node previous = L->cursor->previous;
        if( L->length <= 0 || L->indexCursor< 0){
	   printf("List Error: insertBefore() called on empty List");
           exit(1);
        }
        if( L->indexCursor == 0){
            prepend(L,data);
        }
        else{
            element->previous = L->cursor->previous;
            element->next = L->cursor;
            previous->next = element;
            L->cursor->previous= element;
            L->length++;
            L->indexCursor++;
        }
    }

void insertAfter(List L, int data){
    Node element = newNode(data);
        Node next = L->cursor->next;
        if( L->length <= 0 || L->indexCursor < 0){
            printf("List Error: insertafter() called on empty List");
            exit(1);
        }
        if( L->indexCursor == L->length - 1){
            append(L,data);
        }

        else{
            element->previous = L->cursor;
            element->next = L->cursor->next;
            next->previous = element;
            L->cursor->next= element;
            L->length++;
            L->indexCursor++;
        }
    }

void deleteFront(List L){
        if(L->length <= 0){
            printf("List Error: cannot call front() on empty List\n");
             exit(1);
        }
        else{
	    Node N = L->front;
            L->front = L->front->next;
            L->length--;
	    freeNode(&N);
	    
 }
    }

void deleteBack(List L){
        if(L->length <= 0){
            printf("List Error: cannot call front() on empty List\n");
             exit(1);
        }

        else{
            L->cursor = L->back;
            L->back = L->back->previous;
            L->length--;
        }
    }
           

void delete(List L){
        if(L->length <= 0 || L->indexCursor< 0){
            printf("List Error: cannot call front() on empty List\n");
            exit(1);
      }
      else{
          L->cursor->previous->next = L->cursor->next;
          L->cursor->next = L->cursor->previous;
          L->cursor = NULL;
          L->indexCursor = -1;
          L->length--;
                        }
   }

void printList(FILE* out, List L){
        Node print = L->front;
        while( print != NULL){
            fprintf(out,"%d ", print->data);
            print = print->next;
        }
   }

List copyList(List L){
 List C = newList();
 for(Node T = L->front; T != NULL; T=T->next){
    append(C,T->data);
 }
 return C;
}


